src/coreclr/vm/assembly.cpp

```diff
@@ -643,13 +674,13 @@ Assembly *Assembly::CreateDynamic(AppDomain *pDomain, CreateDynamicAssemblyArgs
         {
             GCX_PREEMP();
             // Assembly::Create will call SuppressRelease on the NewHolder that holds the LoaderAllocator when it transfers ownership
-            pAssem = Assembly::Create(pDomain, pFile, pDomainAssembly->GetDebuggerInfoBits(), args->access & ASSEMBLY_ACCESS_COLLECT ? TRUE : FALSE, pamTracker, pLoaderAllocator);
+            pAssem = Assembly::Create(pDomain, pFile, pDomainAssembly->GetDebuggerInfoBits(), ((args->access & ASSEMBLY_ACCESS_COLLECT) | pLoaderAllocator->IsCollectible()) ? TRUE : FALSE, pamTracker, pLoaderAllocator);
 
             ReflectionModule* pModule = (ReflectionModule*) pAssem->GetManifestModule();
             pModule->SetCreatingAssembly( pCallerAssembly );
 
 
-            if ((args->access & ASSEMBLY_ACCESS_COLLECT) != 0)
+            if ((args->access & ASSEMBLY_ACCESS_COLLECT) & ~usesAllocatorFromBinderContext)
```

Comment:
I would not check for the `ASSEMBLY_ACCESS_COLLECT` at all, just for the new boolean. Same below.

Reply:
Fixed